delegation/delegated.gno

```diff
@@ -0,0 +1,252 @@
+		// if not explicitly delegated, return owner
+		return owner
+	}
+
+	return delegatee
+}
+
+func Delegate(owner, to std.Address) {
+	if gDelegatedToken == nil {
+		panic("DelegatedToken not initialized")
+	}
+
+	currDelegatee := GetDelegatedTo(owner)
+
+	// if the new delegation is different from the current one
+	if currDelegatee != to {
+		gDelegatedToken.delegatedTo[owner] = to
+
+		std.Emit(
+			"Delegation",
+			"from", owner.String(),
+			"to", to.String(),
+		)
+	}
+}
+
+// Deposit transfers the approved amount of the staked token to this contract
+// and mints an GRC20 representing the staked amount.
+func Deposit() {
+	if gDelegatedToken == nil {
+		panic("DelegatedToken not initialized")
+	}
+
+	caller := std.GetOrigCaller()
+	token := gDelegatedToken.token
+
+	// get the approved amount
+	allowance := token.Allowance(caller, std.CurrentRealm().Addr())
+	if allowance == 0 {
+		panic("no allowance")
+	}
+
+	err := token.TransferFrom(caller, std.GetOrigPkgAddr(), allowance)
+	if err != nil {
+		panic("transfer failed: " + err.Error())
+	}
+
+	// mint equivalent amount of delegated tokens
+	amount := u256.NewUint(allowance)
+	mintDelegatedTokens(caller, amount)
+
+	std.Emit(
+		"Deposit",
+		"from", caller.String(),
+		"amount", amount.ToString(),
+	)
+}
+
+// DepositAmount deposites a specific amount of tokens
+func DepositAmount(amount *u256.Uint) {
+	if gDelegatedToken == nil {
+		panic("DelegatedToken not initialized")
+	}
+
+	caller := std.GetOrigCaller()
+	token := gDelegatedToken.token
+
+	// check allowance
+	allowance := token.Allowance(caller, std.GetOrigPkgAddr())
+	if u256.NewUint(allowance).Cmp(amount) < 0 {
+		panic("insufficient allowance")
+	}
+
+	// transfer tokens from caller to this contract
+	err := token.TransferFrom(caller, std.GetOrigPkgAddr(), amount.Uint64())
+	if err != nil {
+		panic("transfer failed: " + err.Error())
+	}
+
+	// mint equivalent amount of delegated tokens
+	mintDelegatedTokens(caller, amount)
+
+	std.Emit(
+		"DepositAmount",
+		"from", caller.String(),
+		"amount", amount.ToString(),
+	)
+}
+
+// Withdraw withdraws the entire staked balance from the contract for the caller
+func Withdraw() {
+	if gDelegatedToken == nil {
+		panic("DelegatedToken not initialized")
+	}
+
+	caller := std.GetOrigCaller()
+	balance := BalanceOf(caller)
+
+	if balance.IsZero() {
+		panic("no balance to withdraw")
+	}
+
+	// burn the delegated tokens
+	gDelegatedToken.balances[caller] = u256.Zero()
+	gDelegatedToken.totalSupply = u256.Zero().Sub(gDelegatedToken.totalSupply, balance)
+
+	// transfer the original tokens back to the caller
+	token := gDelegatedToken.token
+	err := token.Transfer(caller, balance.Uint64())
+	if err != nil {
+		panic("transfer failed: " + err.Error())
+	}
+
+	std.Emit(
+		"Withdrawal",
+		"from", caller.String(),
+		"amount", balance.ToString(),
+	)
+}
+
+// WithdrawAmount withdraws specified amount of token
+// from the contract for the caller
+func WithdrawAmount(amount *u256.Uint) {
```

Comment:
Seems like user has to call this function directly.
If it is make amount parameter to string type then convert to u256

Reply:
https://github.com/gnoswap-labs/governance/pull/9/commits/681e41d9529d86d8ee8adacbe9d59a148474fbec